### ISSUE TITLE
pytorch 0.4 support

### ISSUE THREAD

**Issue by sunlightsgy** [IBM/adversarial-robustness-toolbox#12]
I think the pytorch version is >0.4, but I get the following error

```
======================================================================
ERROR: test_ptclassifier (attacks.carlini_unittest.TestCarliniL2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data1/helend/lucasgyshen/adversarial-robustness-toolbox/art/attacks/carlini_unittest.py", line 268, in test_ptclassifier
    ptc.fit(x_train, y_train, batch_size=batch_size, nb_epochs=1)
  File "/data1/helend/lucasgyshen/adversarial-robustness-toolbox/art/classifiers/pytorch.py", line 143, in fit
    (_, m_batch) = self._model(i_batch)
  File "/data1/helend/anaconda3/envs/lucasgyshen/lib/python3.5/site-packages/torch/nn/modules/module.py", line 477, in __call__
    result = self.forward(*input, **kwargs)
  File "/data1/helend/lucasgyshen/adversarial-robustness-toolbox/art/attacks/carlini_unittest.py", line 46, in forward
    x = self.pool(F.relu(self.conv(x)))
  File "/data1/helend/anaconda3/envs/lucasgyshen/lib/python3.5/site-packages/torch/nn/modules/module.py", line 477, in __call__
    result = self.forward(*input, **kwargs)
  File "/data1/helend/anaconda3/envs/lucasgyshen/lib/python3.5/site-packages/torch/nn/modules/conv.py", line 301, in forward
    self.padding, self.dilation, self.groups)
RuntimeError: Expected object of type torch.FloatTensor but found type torch.cuda.FloatTensor for argument #2 'weight'

```

I think it is because of the new feature of pytorch 0.4. Can you fix it? Thank you.

**Comment by ririnicolae** [comment]
Thanks for letting us know! We'll try to support PyTorch 0.4 soon.

**Comment by samuelemarro** [comment]
I think it's because of PyTorch's `from_numpy()` function, which casts the numpy array to a CPU tensor. If the model is running on CUDA, which happens if PyTorch detects a CUDA-compatible GPU:
```
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
self._model.to(device)
```
then there's a mismatch (`torch.FloatTensor` vs `torch.cuda.FloatTensor`).
The solution is to cast CPU tensors to CUDA tensors with `.to(device)` and convert them to CPU tensors with `.cpu()`. before converting them to numpy arrays. If the environment does not support CUDA, both `.to()` and `.cpu()` do nothing.

**Comment by ririnicolae** [comment]
Thanks, @samuelemarro! I'll try your solution and see if it works for both PyTorch 0.3 and 0.4.

**Comment by marcociccone** [comment]
Hi! Any update on this? Thanks

**Comment by ririnicolae** [comment]
I just pushed a fix (commit 51a001c). @sunlightsgy, @marcociccone, let me know if the issue is solved for you. @samuelemarro I used your solution, thanks for that!
